quantinuum_qircheck/qircheck.py

```diff
@@ -22,6 +22,38 @@
 import pyqir as pq
 
 
+class _cycle_check:
+    # checks for cycles in the CFG
+
+    def __init__(self) -> None:
+        # gray nodes/blocks
+        # marks all blocks that are currently traversed
+        self.current_blocks: set[pq.BasicBlock] = set()
+
+        # black nodes/blocks
+        # marks all blocks that have been traversed
+        self.visited_blocks: set[pq.BasicBlock] = set()
+
+    def check_for_cycles(self, block: pq.BasicBlock) -> None:
+        # checks if the given block is part of a cycle in the CFG
+        # this implements a DFS search and it fails when a backedge is found
+
+        self.current_blocks.add(block)
+        # loop over all the adjacent blocks
+        for instr in block.instructions:
+            if instr.opcode == pq.Opcode.BR or instr.opcode == pq.Opcode.INDIRECT_BR:
+                for x in instr.successors:
+                    if (x not in self.current_blocks) and (
+                        x not in self.visited_blocks
+                    ):
+                        self.check_for_cycles(x)
+                    if x in self.current_blocks:
+                        raise ValueError("Found loop in CFG")
```

Comment:
Can these two `if` blocks be swapped (which would simplify the condition on the one that's currently first), or is the order important?